components/pluralrules/src/rules/ast.rs

```diff
@@ -0,0 +1,314 @@
+/// };
+///
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub struct Relation {
+    pub expression: Expression,
+    pub operator: Operator,
+    pub range_list: RangeList,
+}
+
+/// An enum of Relation operators for plural rules.
+///
+/// Each Operator enumeration belongs to the corresponding symbolic operators:
+///
+/// | Enum Operator | Symbolic Operator |
+/// | - | - |
+/// | In | "in" |
+/// | NotIn | "not in" |
+/// | Within | "within" |
+/// | NotWithin | "not within" |
+/// | Is | "is" |
+/// | IsNot | "is not" |
+/// | Eq | "=" |
+/// | NotEq | "!=" |
+///
+#[derive(Debug, Clone, PartialEq)]
+pub enum Operator {
+    In,
+    NotIn,
+    Within,
+    NotWithin,
+    Is,
+    IsNot,
+    Eq,
+    NotEq,
+}
+
+/// An incomplete AST representation of a plural rule. Comprises an Operand and an optional Modulo.
+///
+/// # Examples
+///
+/// All AST nodes can be built explicitly, as seen in the example. However, due to its complexity, it is preferred to build the AST using the parse_plural_rule function.
+///
+/// ```text
+/// "i % 100"
+/// ```
+///
+/// Can be represented by the AST:
+///
+/// ```
+/// use icu_pluralrules::rules::ast::*;
+///
+/// Expression {
+///     operand: Operand::I,
+///     modulus: Some(Value(100)),
+/// };
+///
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub struct Expression {
+    pub operand: Operand,
+    pub modulus: Option<Value>,
+}
+
+/// An incomplete AST representation of a plural rule. Comprises a char.
+///
+/// # Examples
+///
+/// All AST nodes can be built explicitly, as seen in the example. However, due to its complexity, it is preferred to build the AST using the parse_plural_rule function.
+///
+/// ```text
+/// "i"
+/// ```
+///
+/// Can be represented by the AST:
+///
+/// ```
+/// use icu_pluralrules::rules::ast::Operand;
+///
+/// Operand::I;
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub enum Operand {
+    /// Absolute value of input
+    N,
+    /// Integer value of input
+    I,
+    /// Number of visible fraction digits with trailing zeros
+    V,
+    /// Number of visible fraction digits without trailing zeros
+    W,
+    /// Visible fraction digits with trailing zeros
+    F,
+    /// Visible fraction digits without trailing zeros
+    T,
+}
+
+/// An incomplete AST representation of a plural rule. Comprises a vector of RangeListItems.
+///
+/// # Examples
+///
+/// All AST nodes can be built explicitly, as seen in the example. However, due to its complexity, it is preferred to build the AST using the parse_plural_rule function.
+///
+/// ```text
+/// "5, 7, 9"
+/// ```
+///
+/// Can be represented by the AST:
+///
+/// ```
+/// use icu_pluralrules::rules::ast::*;
+///
+/// RangeList(Box::new([
+///     RangeListItem::Value(Value(5)),
+///     RangeListItem::Value(Value(7)),
+///     RangeListItem::Value(Value(9)),
+/// ]));
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub struct RangeList(pub Box<[RangeListItem]>);
```

Comment:
optional: based on previous discussions about UnicodeSet, I think RangeList is a set of integers in the same way that UnicodeSet is a set of characters.  A range just happens to be compact way of representing certain set instances, but not all set instances are a range.  So maybe rename to `MatchSet`, since it's a set of values (integers) to match against?

Reply:
I followed the EBNF naming scheme here: https://unicode.org/reports/tr35/tr35-numbers.html#Plural_rules_syntax

Do you think we should diverge?

Reply:
My vote is to stick with UTS 35 terminology.

Reply:
Yeah, sticking with UTS 35 makes sense, so keep as is... I wish it were less confusing, but I can't say that I could come up with anything better, anyways.